File: lld/concurrency/python/concurrency.py
"""Producer-consumer worker pool LLD primitive — Python reference implementation.

Multiple producer threads push items into a bounded queue.Queue; a fixed
pool of worker threads drains it concurrently. See ../README.md for the
design writeup.
"""
from __future__ import annotations

import queue
import threading
from dataclasses import dataclass
from typing import Callable


@dataclass(frozen=True)
class Item:
    producer_id: int
    seq: int

    @property
    def id(self) -> str:
        return f"p{self.producer_id}-{self.seq}"


_POISON = Item(-1, -1)


class BoundedPipeline:
    """Runs producers and a worker pool over a bounded queue.Queue.

    Producers push items until exhausted; one poison-pill sentinel per
    worker is enqueued afterwards so every worker thread exits cleanly.
    """

    def __init__(self, num_workers: int, buffer_size: int):
        self.num_workers = num_workers
        self.buffer_size = buffer_size

    def run(self, num_producers: int, items_per_producer: int, handle: Callable[[Item], None]) -> None:
        buffer: "queue.Queue[Item]" = queue.Queue(maxsize=self.buffer_size)

        def produce(producer_id: int) -> None:
            for seq in range(items_per_producer):
                buffer.put(Item(producer_id, seq))

        def consume() -> None:
            while True:
                item = buffer.get()
                if item is _POISON:
                    buffer.task_done()
                    return
                handle(item)
                buffer.task_done()

        producers = [threading.Thread(target=produce, args=(pid,)) for pid in range(num_producers)]
        workers = [threading.Thread(target=consume) for _ in range(self.num_workers)]

        for w in workers:
            w.start()
        for p in producers:
            p.start()
        for p in producers:
            p.join()

        for _ in workers:
            buffer.put(_POISON)
        for w in workers:
            w.join()


def _demo() -> None:
    consumed: set[str] = set()
    lock = threading.Lock()

    def handle(item: Item) -> None:
        with lock:
            consumed.add(item.id)

    pipeline = BoundedPipeline(num_workers=4, buffer_size=8)
    pipeline.run(num_producers=3, items_per_producer=20, handle=handle)
    print(f"Consumed {len(consumed)} unique items via producer-consumer pipeline")


if __name__ == "__main__":
    _demo()

File: lld/concurrency/go/concurrency.go
// Package concurrency demonstrates the core concurrency primitives used in
// LLD interviews: goroutines, a bounded channel as the shared buffer, a
// worker-pool of consumers, and sync.WaitGroup for lifecycle coordination.
package concurrency

import (
	"fmt"
	"sync"
)

// Item is a unit of work flowing from producers to workers.
type Item struct {
	ProducerID int
	Seq        int
}

// ID uniquely identifies an item across all producers so consumers can
// dedupe/verify delivery.
func (i Item) ID() string {
	return fmt.Sprintf("p%d-%d", i.ProducerID, i.Seq)
}

// Pool runs a bounded producer-consumer pipeline: numProducers goroutines
// each emit itemsPerProducer items into a buffered channel of size
// bufferSize (the bounded buffer), and numWorkers goroutines drain it
// concurrently, invoking handle for every item they receive.
type Pool struct {
	NumWorkers int
	BufferSize int
}

// Run starts producers and workers, blocks until every item has been
// produced and consumed, then returns. handle is called once per item from
// whichever worker goroutine received it, so callers must make it
// concurrency-safe themselves (e.g. guard shared state with a mutex).
func (p *Pool) Run(numProducers, itemsPerProducer int, handle func(Item)) {
	buffer := make(chan Item, p.BufferSize)

	var producers sync.WaitGroup
	producers.Add(numProducers)
	for pid := 0; pid < numProducers; pid++ {
		go func(pid int) {
			defer producers.Done()
			for seq := 0; seq < itemsPerProducer; seq++ {
				buffer <- Item{ProducerID: pid, Seq: seq}
			}
		}(pid)
	}

	var workers sync.WaitGroup
	workers.Add(p.NumWorkers)
	for w := 0; w < p.NumWorkers; w++ {
		go func() {
			defer workers.Done()
			for item := range buffer {
				handle(item)
			}
		}()
	}

	// Closing the channel only after all producers finish signals workers
	// to drain the remaining buffered items and exit the range loop.
	producers.Wait()
	close(buffer)
	workers.Wait()
}

// SafeSet is a mutex-guarded string set used to collect consumed item IDs
// concurrently without losing updates (the map itself is not goroutine-safe).
type SafeSet struct {
	mu   sync.Mutex
	seen map[string]int
}

func NewSafeSet() *SafeSet {
	return &SafeSet{seen: make(map[string]int)}
}

func (s *SafeSet) Add(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.seen[id]++
}

func (s *SafeSet) Snapshot() map[string]int {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make(map[string]int, len(s.seen))
	for k, v := range s.seen {
		out[k] = v
	}
	return out
}
